pennylane/devices/qubit/simulate.py

```diff
@@ -630,39 +659,229 @@ def find_next_mcm(circuit):
     return circuit_base, circuit_next, op
 
 
+def prepend_state_prep(circuit, state, interface, wires):
+    """Prepend a ``StatePrep`` operation with the prescribed ``wires`` to the circuit.
+
+    ``get_final_state`` executes a circuit on a subset of wires found in operations
+    or measurements. This function makes sure that an initial state with the correct size is created
+    on the first invocation of ``simulate_tree_mcm``. ``wires`` should be the wires attribute
+    of the original circuit (which included all wires)."""
+    if len(circuit) > 0 and isinstance(circuit[0], qml.operation.StatePrepBase):
+        return circuit
+    state = (
+        create_initial_state(wires, None, like=INTERFACE_TO_LIKE[interface])
+        if state is None
+        else state
+    )
+    return qml.tape.QuantumScript(
+        [qml.StatePrep(state.ravel(), wires=wires, validate_norm=False)] + circuit.operations,
+        circuit.measurements,
+        shots=circuit.shots,
+        trainable_params=circuit.trainable_params,
+    )
+
+
+def insert_mcms(circuit, results, mid_measurements):
+    """Inserts terminal measurements of MCMs if the circuit is evaluated in analytic mode."""
+    if circuit.shots:
+        return results
+    if not any(m.mv for m in circuit.measurements):
+        return results
+    results = list(results)
+    new_results = []
+    mid_measurements = dict((k, qml.math.array([[v]])) for k, v in mid_measurements.items())
+    for m in circuit.measurements:
+        if m.mv:
+            new_results.append(gather_mcm(m, mid_measurements, qml.math.array([[True]])))
```

Comment:
This makes sense, but it seems a bit hacky to use a function meant to be used with finite shots.

Reply:
Yes, it is a bit like using the integral mean value theorem when I know what the mean value is.